### PR TITLE
doc(Trial): Better indication of what to do with the Metric name for report

### LINKED ISSUE #211: [UX] Better indication of what to do with the `Metric` name
Right now, there's an unclear connection between the `Metric` you need to give to the `Optimizer` and the fact you need to return this as part of the `Trial.Report`, i.e. with `trial.success(metric_name=val)`

We can't make these live together as the `Metric` needs to go to the `Optimizer` and there's no way to hint this with type checking.

Two things to do:
---

* Improve the error message with a hint on why this might happen and what to do.
```python
ValueError: Cannot report success without self.metrics=[Metric(name='poly_metric', minimize=False, bounds=None)]. Please provide a value for the metric.
```
The suggestion is to explicitly state the solutions, (replacing with f_strings below ofc):

```python
ValueError: Cannot report success without self.metrics=[Metric(name='poly_metric', minimize=False, bounds=None)]. Please provide a value for the metric.

Please provide `{metric.name}` as `trial.success({metric.name}=value)` or rename your metric to
`Metric(name="{provided_key}", minimize={metric.minimize}, bounds={metric.bounds})`
```

Might be a bit laborious to account for multiple metrics as well but it's worth the explicit hinting

* Improve the documentation by explicitly stating this with an annotated block like `!!! tip`

Quick Reproduce
---
```python
from amltk.optimization import Metric, Trial

m = Metric("acc", minimize=False, bounds=(0, 1))

t = Trial(name="dummy", metrics=[m], config={})

with t.begin():
    pass

# NOTE: User gave `accuracy` instead of `acc`
report = t.success(accuracy=0.5)

# And multiobjective, what to give as good error message?
# Maybe we can excpect the people doing multi-objective already
# know the basics and can figure it out. Might not need
# to explicitly support it.


loss = Metric("loss", minimize=True, bounds=None)

t2 = Trial(name="dummy", metrics=[m, loss], config={})

with t2.begin():
    pass


report = t2.success(acc=0.5, los=0.5)  # acc and los (one s)

```

### PR DESCRIPTION
**Pull request by vladislavalerievich**
[[UX] Better indication of what to do with the Metric name #211](https://github.com/automl/amltk/issues/211)

* Closes #211

### CODE REVIEW

File: src/amltk/optimization/trial.py [hunk 1]
```diff
@@ -130,13 +130,53 @@ def target_function(trial: Trial) -> Trial.Report:
...
     [`.name`][amltk.optimization.Trial.name] given the optimization run,
-    a candidate [`.config`][amltk.optimization.Trial.config]' to evaluate,
+    a candidate [`.config`][amltk.optimization.Trial.config] to evaluate,
     a possible [`.seed`][amltk.optimization.Trial.seed] to use,
-    and an [`.info`][amltk.optimization.Trial.info] object which is the optimizer
+    and an [`.info`][amltk.optimization.Trial.info] object, which is the optimizer
     specific information, if required by you.
 
+    !!! tip
+    When using the `success` method, make sure to provide values for all metrics
```
Comment (eddiebergman) [inline src/amltk/optimization/trial.py:141]: You'll need to add a blank line between these and also indent it one level. Again, try render the docs to see if it renders correctly

**Comment by eddiebergman** [comment]
You can close an issue automatically using this syntax in github, you don't need to hyperlink it.

```
* Closes #216 
```

By doing the above, it'll automatically close it once it's merged.

Lastly, can you name the PR's in the same way as the commit messages, i.e. `refactor: xyz`. This is because when it's merged into `main`, the PR title will be the commit message

File: src/amltk/optimization/trial.py [hunk 2]
```diff
@@ -130,13 +130,53 @@ def target_function(trial: Trial) -> Trial.Report:
...
+        config={"param": 42},
+        metrics=[Metric(name="poly_metric", minimize=False)]
+    )
+
+    # Incorrect usage (will raise a ValueError)
+    report = trial.success(poly_metric=0.95)
+
+    # Correct usage
+    report = trial.success(provided_key=0.95)
+    trial.metrics = [Metric(name="provided_key", minimize=False, bounds=None)]
```
Comment (eddiebergman) [inline src/amltk/optimization/trial.py:177]: You should neveer really change the metrics for the trial, they're set by the Optimizer. I'll update the docs to reflect this

**Review by eddiebergman — CHANGES_REQUESTED** [review]

**Comment by eddiebergman** [comment]
So I pushed some more things. Namely I added error checks to ensure that the metrics are indeed unique (as you specified in the docs) and then also ensure the metric name is a valid python identifier

**Comment by eddiebergman** [comment]
Can you do another review just to see what's changed?

**Comment by vladislavalerievich** [comment]
> Can you do another review just to see what's changed?

Thank you for adding additional checks.

**Pull request merged**